### PR TITLE
Chore: Upgrade typescript to 3.7

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Upgrades typescript to the latest version.

### CODE REVIEW

File: packages/grafana-ui/src/slate-plugins/suggestions.tsx [hunk 1]
```diff
@@ -6,14 +6,7 @@ import { Editor as CoreEditor } from 'slate';
...
 import TOKEN_MARK from './slate-prism/TOKEN_MARK';
-import {
-  makeFragment,
-  TypeaheadOutput,
-  CompletionItem,
-  TypeaheadInput,
-  SuggestionsState,
-  CompletionItemGroup,
-} from '..';
+import { makeFragment, TypeaheadOutput, CompletionItem, TypeaheadInput, SuggestionsState } from '..';
```
Comment (dprokop) [inline packages/grafana-ui/src/slate-plugins/suggestions.tsx]: Could you please update this import to be from the accurate path rather than `..`? We also need to introduce linter for such imports as they are problematic for Storybook.

File: public/app/features/plugins/datasource_srv.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import { GrafanaRootScope } from 'app/routes/GrafanaCtrl';
 import { expressionDatasource } from 'app/features/expressions/ExpressionDatasource';
 
 export class DatasourceSrv implements DataSourceService {
-  datasources: { [name: string]: DataSourceApi };
+  datasources: Record<string, DataSourceApi>;
```
Comment (dprokop) [inline public/app/features/plugins/datasource_srv.ts]: Hm, never seen `Record` before. Is this a new thing introduced in TS3.7? We also have wildly used  `KeyValue` type in grafana/data which could probably be replaced with `Record`

File: packages/grafana-data/src/types/datasource.ts [hunk 3]
```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
...
-import { RawTimeRange, TimeRange } from './time';
+import { RawTimeRange, TimeRange, AbsoluteTimeRange } from './time';
 import { ScopedVars } from './ScopedVars';
 
 export interface DataSourcePluginOptionsEditorProps<JSONData = DataSourceJsonData, SecureJSONData = {}> {
   options: DataSourceSettings<JSONData, SecureJSONData>;
   onOptionsChange: (options: DataSourceSettings<JSONData, SecureJSONData>) => void;
 }
 
+export type DataSourceQuery<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
```
Comment (dprokop) [inline packages/grafana-data/src/types/datasource.ts]: @kaydelaney can you say something more about these changes? Can't see how they fit into TS upgrade.

cc @ryantxu

File: packages/grafana-data/src/types/datasource.ts [hunk 4]
```diff
@@ -284,6 +298,7 @@ export interface ExploreQueryFieldProps<
 > extends QueryEditorProps<DSType, TQuery, TOptions> {
   history: any[];
   onBlur?: () => void;
+  absoluteRange?: AbsoluteTimeRange;
```
Comment (dprokop) [inline packages/grafana-data/src/types/datasource.ts]: Why is this added here? 

cc @ryantxu

File: packages/grafana-ui/src/components/Chart/Tooltip.test.tsx [hunk 5]
```diff
@@ -12,7 +12,7 @@ describe('Chart Tooltip', () => {
       // We need to mock getBoundingClientRect to return what DOM would actually return
       // when measuring tooltip container (wrapper with padding and content inside)
       Element.prototype.getBoundingClientRect = jest.fn(() => {
-        return { width: 100, height: 100, top: 0, left: 0, bottom: 0, right: 0 };
+        return { width: 100, height: 100, top: 0, left: 0, bottom: 0, right: 0 } as any;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Chart/Tooltip.test.tsx]: `as ClientRect | DOMRect` ?

**Comment by dprokop** [comment]
Added @ryantxu as I see some changes he might be interested in

File: packages/grafana-ui/src/slate-plugins/suggestions.tsx (hunk 1 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/slate-plugins/suggestions.tsx]: Sure

File: public/app/features/plugins/datasource_srv.ts (hunk 2 above)
Comment (kaydelaney) [reply public/app/features/plugins/datasource_srv.ts]: It's existed for a while as far as I know (https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt). Mostly a personal preference thing but I think it reads a bit better than `{ [key: string]: DataSourceApi }`

File: packages/grafana-data/src/types/datasource.ts (hunk 3 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/datasource.ts]: Sure. Typescript 3.7 is somewhat more strict with its typechecking in certain scenarios, which caused a type error when doing the following
```typescript
const plugin = new DataSourcePlugin(lokiDatasource);
```
with the error
```
"Argument of type 'typeof LokiDatasource' is not assignable to parameter of type 'DataSourceConstructor<DataSourceApi<DataQuery, LokiOptions>, DataQuery, LokiOptions>'."
```
So in order for the correct types to be properly inferred the helper types above were introduced to extract the `TQuery` and `TOptions` types of a type extending `DataSourceApi<TQuery, TOptions>`.

File: packages/grafana-data/src/types/datasource.ts (hunk 4 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/datasource.ts]: An `absoluteRange` prop is passed to `QueryField` components at some points (see QueryRow.tsx, line 168), which causes an error in typescript 3.7 since `absoluteRange` wasn't a member of the props interface.

Full error:
```
Type '{ datasource: DataSourceApi<DataQuery, DataSourceJsonData>; query: DataQuery; history: HistoryItem<DataQuery>[]; ... 5 more ...; absoluteRange: AbsoluteTimeRange; }' is not assignable to type 'IntrinsicAttributes & ExploreQueryFieldProps<DataSourceApi<DataQuery, DataSourceJsonData>, DataQuery, DataSourceJsonData> & { ...; }'.
  Property 'absoluteRange' does not exist on type 'IntrinsicAttributes & ExploreQueryFieldProps<DataSourceApi<DataQuery, DataSourceJsonData>, DataQuery, DataSourceJsonData> & { ...; }'.ts(2322)
```

File: packages/grafana-ui/src/components/Chart/Tooltip.test.tsx (hunk 5 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Chart/Tooltip.test.tsx]: good idea

**Review by davkal — APPROVED** [review]

File: packages/grafana-data/src/types/datasource.ts [hunk 6]
```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
...
+  ? TQuery
+  : never;
+
+export type DataSourceOptions<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
+  infer _TQuery,
+  infer TOptions
+>
+  ? TOptions
+  : never;
+
```
Comment (ryantxu) [inline packages/grafana-data/src/types/datasource.ts]: **head... exploding**

I'll trust you on this one :)  The nested typing on the datasources getting super hard to follow - I think that is just the nature of the structure, but is something we should try to clean up if possible

**Review by ryantxu — APPROVED** [review]
Looks good -- the DataSourceAPI typing is complex, but that is not new :(

File: packages/grafana-data/src/types/datasource.ts (hunk 6 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/datasource.ts]: Agreed

**Review by dprokop — APPROVED** [review]
@kaydelaney thanks for explaining all the things, great stuff!
